tensorzero-internal/tests/e2e/providers/common.rs

```diff
@@ -10315,3 +10325,194 @@ pub async fn test_multi_turn_parallel_tool_use_streaming_inference_request_with_
         }
     }
 }
+
+pub async fn test_json_mode_off_inference_request_with_provider(provider: E2ETestProvider) {
+    // Some providers don't support configuring JSON mode,
+    // so we'll skip the test for them
+    if provider.model_provider_name == "sglang"
+        || provider.model_provider_name == "tgi"
+        || provider.model_provider_name == "vllm"
+    {
+        return;
+    }
+
+    let episode_id = Uuid::now_v7();
+
+    let payload = json!({
+        "function_name": "json_success",
+        "variant_name": provider.variant_name,
+        "episode_id": episode_id,
+        "input":
+            {
+               "system": {"assistant_name": "AskJeeves"},
+               "messages": [
+                   {
+                       "role": "user",
+                       "content": {"country": "Japan"}
+                   }
+               ]
+            },
+        "params": {
+            "chat_completion": {
+                "json_mode": "off",
+            }
+        },
+        "stream": false,
+    });
+
+    let response = Client::new()
+        .post(get_gateway_endpoint("/inference"))
+        .json(&payload)
+        .send()
+        .await
+        .unwrap();
+
+    // Check that the API response is ok
+    assert_eq!(response.status(), StatusCode::OK);
+    let response_json = response.json::<Value>().await.unwrap();
+
+    println!("API response: {response_json:#?}");
+
+    // Check basic response structure
+    let output = response_json.get("output").unwrap().as_object().unwrap();
+    let parsed = output.get("parsed").unwrap().as_object().unwrap();
+    let answer = parsed.get("answer").unwrap().as_str().unwrap();
+
+    // Assert that Tokyo is the answer
+    assert_eq!(answer, "Tokyo");
+    let raw = output.get("raw").unwrap().as_str().unwrap();
+    assert_eq!(raw, "{\n    \"answer\": \"Tokyo\"\n}");
+
+    // The result shouldn't be valid JSON when json_mode is off
+    let is_valid_json = serde_json::from_str::<Value>(raw).is_ok();
+    if is_valid_json {
+        // Some providers might still return JSON even with json_mode off
+        // This is not a failure, just something to be aware of
+        println!("WARNING: Provider returned valid JSON despite json_mode being off");
+    }
```

Comment:
Let's ask the providers to generate something that is not JSON (for a JSON function with output schema) and assert that they do not return valid JSON. Otherwise this test will likely not catch any regressions.

Reply:
See if sending a user message along the lines of "Before your JSON, write ..." will do it. If not, creating an extra function is fine. Thanks!